File: apps/ios/Pebbles/Features/Path/Components/RippleBadge.swift
import SwiftUI

/// 44×44 ring-and-digit badge representing the user's Ripples level.
/// See `docs/superpowers/specs/2026-05-15-ripples-design.md` and
/// issue #442 for full color/state semantics.
struct RippleBadge: View {
    let level: Int
    let activeToday: Bool

    @Environment(\.colorScheme) private var colorScheme

    private var clampedLevel: Int { min(max(level, 0), 6) }

    private var digitColor: Color {
        switch (colorScheme, activeToday) {
        case (.dark, true):   return .pebblesSurface
        case (.light, true):  return .pebblesForeground
        case (.dark, false),
             (.light, false): return .pebblesMutedForeground
        @unknown default:     return .pebblesForeground
        }
    }

    private func tone(forStroke id: Int) -> RippleStrokeTone {
        rippleStrokeTone(strokeId: id, level: clampedLevel, activeToday: activeToday)
    }

    var body: some View {
        ZStack {
            // Draw outermost first so inner rings paint on top.
            RippleStroke6()
                .stroke(tone(forStroke: 6).color, style: stroke)
                .opacity(0.33)
            RippleStroke5()
                .stroke(tone(forStroke: 5).color, style: stroke)
                .opacity(0.33)
            RippleStroke4()
                .stroke(tone(forStroke: 4).color, style: stroke)
                .opacity(0.33)
            RippleStroke3()
                .stroke(tone(forStroke: 3).color, style: stroke)
                .opacity(0.66)
            RippleStroke2()
                .stroke(tone(forStroke: 2).color, style: stroke)
                .opacity(0.66)
            RippleStroke1()
                .stroke(tone(forStroke: 1).color, style: stroke)

            Text(verbatim: "\(clampedLevel)")
                .font(.system(size: 12, weight: .bold, design: .rounded))
                .foregroundStyle(digitColor)
        }
        .frame(width: 44, height: 44)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(accessibilityLabel)
    }

    private var stroke: StrokeStyle {
        StrokeStyle(lineWidth: 2, lineCap: .round)
    }

    private var accessibilityLabel: LocalizedStringResource {
        activeToday
            ? LocalizedStringResource("Ripple level \(clampedLevel), active today")
            : LocalizedStringResource("Ripple level \(clampedLevel), inactive today")
    }
}

#Preview("All states — light") {
    RipplePreviewGrid()
        .preferredColorScheme(.light)
}

#Preview("All states — dark") {
    RipplePreviewGrid()
        .preferredColorScheme(.dark)
}

private struct RipplePreviewGrid: View {
    var body: some View {
        VStack(spacing: 12) {
            ForEach([true, false], id: \.self) { active in
                HStack(spacing: 8) {
                    Text(verbatim: active ? "active" : "inactive")
                        .font(.caption)
                        .frame(width: 60, alignment: .leading)
                    ForEach(0...6, id: \.self) { level in
                        RippleBadge(level: level, activeToday: active)
                    }
                }
            }
        }
        .padding()
        .background(Color.pebblesPathBackground)
    }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/shared/ripples/RippleBadge.kt
package app.pbbls.android.features.shared.ripples

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.scale
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.unit.dp
import app.pbbls.android.R
import app.pbbls.android.theme.PebblesText
import app.pbbls.android.theme.PebblesTheme
import app.pbbls.android.theme.PebblesTypography

/**
 * 44×44 ring-and-digit badge representing the user's Ripples level — ports
 * iOS `RippleBadge.swift`. See `docs/superpowers/specs/2026-05-15-ripples-design.md`
 * and issue #442 for the full color/state semantics. Outermost rings draw
 * first so inner rings paint on top; ring opacity steps 0.33 / 0.66 / 1.0
 * from outside in.
 */
@Composable
fun RippleBadge(
    level: Int,
    activeToday: Boolean,
    modifier: Modifier = Modifier,
) {
    val clampedLevel = level.coerceIn(0, 6)
    // Resolve the palette bundle in composition; the tone→color mapping is a
    // pure function so the DrawScope lambda can call it.
    val colors = PebblesTheme.colors
    val digitColor = colors.system.foreground
    val label =
        if (activeToday) {
            stringResource(R.string.ripple_badge_active, clampedLevel)
        } else {
            stringResource(R.string.ripple_badge_inactive, clampedLevel)
        }

    Box(
        modifier =
            modifier
                .size(44.dp)
                .clearAndSetSemantics { contentDescription = label },
        contentAlignment = Alignment.Center,
    ) {
        Canvas(modifier = Modifier.fillMaxSize()) {
            val stroke = Stroke(width = 2.dp.toPx(), cap = StrokeCap.Round)
            // Uniform scale from the 44-unit source viewBox into the canvas.
            scale(
                scaleX = size.width / RIPPLE_VIEWBOX_SIDE,
                scaleY = size.height / RIPPLE_VIEWBOX_SIDE,
                pivot = Offset.Zero,
            ) {
                listOf(
                    6 to 0.33f,
                    5 to 0.33f,
                    4 to 0.33f,
                    3 to 0.66f,
                    2 to 0.66f,
                    1 to 1f,
                ).forEach { (id, alpha) ->
                    val tone = rippleStrokeTone(strokeId = id, level = clampedLevel, activeToday = activeToday)
                    drawPath(
                        path = RippleStrokes.byId.getValue(id),
                        color = tone.color(colors),
                        alpha = alpha,
                        style = stroke,
                    )
                }
            }
        }
        // The Box's clearAndSetSemantics carries the combined a11y label.
        PebblesText(
            text = clampedLevel.toString(),
            style = PebblesTypography.captionEmphasized,
            color = digitColor,
        )
    }
}
